iscsi.go

```diff
@@ -0,0 +1,475 @@
+	_ = C.iscsi_set_header_digest(d.Context, C.ISCSI_HEADER_DIGEST_NONE_CRC32C)
+}
+
+func (d *device) Connect() error {
+	d.initializeContext()
+	return retry.Do(func() error {
+		portalStr := C.CString(d.targetPortal)
+		defer C.free(unsafe.Pointer(portalStr))
+		if retval := C.iscsi_full_connect_sync(d.Context, portalStr, C.int(d.targetLun)); retval != 0 {
+			errstr := C.iscsi_get_error(d.Context)
+			// reset the context before retrying.  it seems like some connection
+			// errors leave the context in an inconsistent state that makes it
+			// difficult to reuse
+			d.initializeContext()
+			return fmt.Errorf("iscsi_full_connect_sync: (%d) %s", retval, C.GoString(errstr))
+		}
+		return nil
+	}, retry.Attempts(20), retry.MaxDelay(500*time.Millisecond))
+}
+
+func (d *device) Reconnect() error {
+	if retval := C.iscsi_reconnect_sync(d.Context); retval != 0 {
+		if retval != 0 {
+			return fmt.Errorf("failed to reconnect with: %d", retval)
+		}
+	}
+	return nil
+}
+
+func (d *device) Disconnect() error {
+	defer C.iscsi_destroy_context(d.Context)
+	retval := C.iscsi_logout_sync(d.Context)
+	if retval != 0 {
+		return fmt.Errorf("failed to logout with: %d", retval)
+	}
+	return nil
+}
+
+type Capacity struct {
+	LBA       int
+	BlockSize int
+}
+
+func (d device) ReadCapacity10() (c Capacity, err error) {
+	task := C.iscsi_readcapacity10_sync(d.Context, 0, 0, 0)
+	defer func() {
+		if task != nil {
+			C.scsi_free_scsi_task(task)
+		}
+	}()
+	if task == nil || task.status != C.SCSI_STATUS_GOOD {
+		errstr := C.iscsi_get_error(d.Context)
+		return c, fmt.Errorf("iscsi_readcapacity10_sync: %s", C.GoString(errstr))
+	}
+	readcapacity, err := getReadCapacity10(*task)
+	if err != nil {
+		return c, err
+	}
+	c.BlockSize = int(readcapacity.block_size)
+	c.LBA = int(readcapacity.lba)
+	logger().Debug("ReadCapacity10", slog.Any("capacity", c))
+	return c, nil
+}
+
+func (d device) ReadCapacity16() (c Capacity, err error) {
+	task := C.iscsi_readcapacity16_sync(d.Context, 0)
+	defer func() {
+		if task != nil {
+			C.scsi_free_scsi_task(task)
+		}
+	}()
+	if task == nil || task.status != C.SCSI_STATUS_GOOD {
+		errstr := C.iscsi_get_error(d.Context)
+		return c, fmt.Errorf("iscsi_readcapacity16_sync: %s", C.GoString(errstr))
+	}
+	readcapacity, err := getReadCapacity16(*task)
+	if err != nil {
+		return c, err
+	}
+	c.BlockSize = int(readcapacity.block_length)
+	c.LBA = int(readcapacity.returned_lba)
+	logger().Debug("ReadCapacity16", slog.Any("capacity", c))
+	return c, nil
+}
+
+type Write16 struct {
+	LBA       int
+	Data      []byte
+	BlockSize int
+}
+
+func (d *device) Write16(data Write16) error {
+	logger().Debug("Write16", slog.Any("request", data))
+	carr := []C.uchar(string(data.Data))
+	// TODO: (willgorman) figure out why larger blocksizes cause SCSI_SENSE_ASCQ_INVALID_FIELD_IN_INFORMATION_UNIT
+	task := C.iscsi_write16_sync(d.Context, 0,
+		C.uint64_t(data.LBA), &carr[0], C.uint(len(carr)), C.int(data.BlockSize), 0, 0, 0, 0, 0)
```

Comment:
`&carr[0]` 

Does this mean we're only writing the first byte?

Reply:
No, C expects arrays to be passed as just a pointer to the first element in the array along with a second parameter for the length of the array.